### PR TITLE
Unstack: take type info from eltype of key column

### LINKED ISSUE #1039: Unstack fails when DataFrame column begins with NA
unstack appears to fail when the DataFrame value column begins with a NA.  The reproducer below demonstrates the problem.  Example output follows.

https://gist.github.com/chipkent/83081d8aa2cfee1cbef074ed6a0cb37b

GOOD permutation...start
10×3 DataFrames.DataFrame
│ Row │ S   │ D    │ V    │
├─────┼─────┼──────┼──────┤
│ 1   │ "B" │ "d1" │ 6.0  │
│ 2   │ "B" │ "d3" │ NA   │
│ 3   │ "B" │ "d5" │ 10.0 │
│ 4   │ "A" │ "d4" │ 4.0  │
│ 5   │ "A" │ "d1" │ 1.0  │
│ 6   │ "A" │ "d2" │ NA   │
│ 7   │ "A" │ "d5" │ 5.0  │
│ 8   │ "B" │ "d2" │ 7.0  │
│ 9   │ "A" │ "d3" │ 3.0  │
│ 10  │ "B" │ "d4" │ 9.0  │
GOOD permutation worked
BAD permutation...start
10×3 DataFrames.DataFrame
│ Row │ S   │ D    │ V    │
├─────┼─────┼──────┼──────┤
│ 1   │ "B" │ "d3" │ NA   │
│ 2   │ "B" │ "d1" │ 6.0  │
│ 3   │ "B" │ "d5" │ 10.0 │
│ 4   │ "A" │ "d4" │ 4.0  │
│ 5   │ "A" │ "d1" │ 1.0  │
│ 6   │ "A" │ "d2" │ NA   │
│ 7   │ "A" │ "d5" │ 5.0  │
│ 8   │ "B" │ "d2" │ 7.0  │
│ 9   │ "A" │ "d3" │ 3.0  │
│ 10  │ "B" │ "d4" │ 9.0  │
ERROR: LoadError: MethodError: `convert` has no method matching convert(::Type{DataArrays.NAtype}, ::Float64)
This may have arisen from a call to the constructor DataArrays.NAtype(...),
since type constructors fall back to convert methods.
Closest candidates are:
  call{T}(::Type{T}, ::Any)
  convert{T}(::Type{T}, !Matched::T)
  DataArrays.NAtype()
 in setindex! at /cc/home/chip/.julia/v0.4/DataArrays/src/indexing.jl:208
 in unstack at /cc/home/chip/.julia/v0.4/DataFrames/src/abstractdataframe/reshape.jl:187
 in unstack at /cc/home/chip/.julia/v0.4/DataFrames/src/abstractdataframe/reshape.jl:192

julia> versioninfo()
Julia Version 0.4.6
Commit 2e358ce (2016-06-19 17:16 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

### PR DESCRIPTION
**Pull request by amellnik**
Fixes #1039

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -170,7 +170,7 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
     # TODO make fillNA(type, length)
-    payload = DataFrame(Any[DataArray(fill(valuecol[1], Nrow), fill(true, Nrow)) for i in 1:Ncol], map(@compat(Symbol), keycol.pool))
+    payload = DataFrame(Any[DataArray(eltype(valuecol), (Nrow)) for i in 1:Ncol], map(@compat(Symbol), keycol.pool))
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl]: `(Nrow)` is strictly equivalent to `Nrow` (you'd have to write `(Nrow,)` if you wanted a tuple), so better use the simpler version.

**Comment by nalimilan** [comment]
Thanks!

**Comment by nalimilan** [comment]
Could you add a test for the case which was previously failing?

**Comment by amellnik** [comment]
Whoops, I should have known to include this the first time around.

File: src/abstractdataframe/reshape.jl [hunk 2]
```diff
@@ -170,7 +170,7 @@ function unstack(df::AbstractDataFrame, rowkey::Int, colkey::Int, value::Int)
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
     # TODO make fillNA(type, length)
-    payload = DataFrame(Any[DataArray(fill(valuecol[1], Nrow), fill(true, Nrow)) for i in 1:Ncol], map(@compat(Symbol), keycol.pool))
+    payload = DataFrame(Any[DataArray(eltype(valuecol), Nrow) for i in 1:Ncol], map(@compat(Symbol), keycol.pool))
```
Comment (ararslan) [inline src/abstractdataframe/reshape.jl]: Shouldn't need `@compat(Symbol)`, I think just `Symbol` is fine?

Comment (amellnik) [inline src/abstractdataframe/reshape.jl]: Passes without @compat(Symbol) on 0.5.0-rc3.  I don't have any easy way to check 0.4.6 offline but can push this and see how the build goes.

**Comment by amellnik** [comment]
Two fails on x32 due to RDA.jl -- I think that's still slated to be removed.

**Pull request merged**